In [139]:
import matplotlib.pyplot as plt 
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

In [140]:
from google.colab import files
uploaded = files.upload()

Saving titanic.csv to titanic (2).csv


In [141]:
pima = pd.read_csv("titanic.csv")
pima.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [142]:
pima = pima.dropna()
pima = pima.reset_index()
for i in pima.columns:
  try:
    pima[[i]] = pima[[i]].astype(float).astype(int)
  except:
    pass


Binomial Logistic Regression

In [143]:
x = pima["Age"]
y = pima["Survived"]

In [144]:
model = LogisticRegression()

In [145]:
model.fit(x.values.reshape(-1, 1), y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [146]:
LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [147]:
model.intercept_

array([2.08695658])

In [148]:
model.coef_

array([[-0.03688114]])

In [149]:
model.predict_proba(x.values.reshape(-1, 1))

array([[0.33504082, 0.66495918],
       [0.3108568 , 0.6891432 ],
       [0.47617429, 0.52382571],
       [0.12571027, 0.87428973],
       [0.51303442, 0.48696558],
       [0.30301156, 0.69698844],
       [0.25840317, 0.74159683],
       [0.20001271, 0.79998729],
       [0.43050752, 0.56949248],
       [0.57696275, 0.42303725],
       [0.39476998, 0.60523002],
       [0.26553347, 0.73446653],
       [0.23777298, 0.76222702],
       [0.22466403, 0.77533597],
       [0.40361519, 0.59638481],
       [0.62985721, 0.37014279],
       [0.22466403, 0.77533597],
       [0.21207672, 0.78792328],
       [0.41252351, 0.58747649],
       [0.23115352, 0.76884648],
       [0.28766327, 0.71233673],
       [0.47617429, 0.52382571],
       [0.20001271, 0.79998729],
       [0.32687473, 0.67312527],
       [0.23115352, 0.76884648],
       [0.31881225, 0.68118775],
       [0.21830499, 0.78169501],
       [0.54060858, 0.45939142],
       [0.49459701, 0.50540299],
       [0.43957191, 0.56042809],
       [0.

In [150]:
model.predict(x.values.reshape(-1, 1))

array([1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1])

In [151]:
model.score(x.values.reshape(-1, 1),y)

0.6939890710382514

In [152]:
confusion_matrix(y,model.predict(x.values.reshape(-1, 1)))

array([[ 11,  49],
       [  7, 116]])

In [153]:
print(classification_report(y,model.predict(x.values.reshape(-1, 1))))

              precision    recall  f1-score   support

           0       0.61      0.18      0.28        60
           1       0.70      0.94      0.81       123

    accuracy                           0.69       183
   macro avg       0.66      0.56      0.54       183
weighted avg       0.67      0.69      0.63       183



Multinomial Logistic Regression

In [154]:
pip install sklearn

In [155]:
import sklearn

In [156]:
from sklearn.model_selection import train_test_split

In [157]:
from google.colab import files
uploaded  = files.upload()

Saving ph-data.csv to ph-data (1).csv


In [158]:
mfd = pd.read_csv("ph-data.csv")
mfd.head()
mfd['label'].unique()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [159]:
xb = mfd.iloc[:,:-1]
yb = mfd.iloc[:,-1]

In [160]:
mn_model = LogisticRegression(multi_class = 'multinomial', solver='lbfgs')

In [161]:
xb_train, xb_test,yb_train,yb_test = train_test_split(xb,yb,train_size=0.75,random_state=2)

In [162]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
xb_train = sc.fit_transform(xb_train)


In [163]:
xb_test = sc.transform(xb_test)

In [164]:
mn_model.fit(xb_train,yb_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='multinomial', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [165]:
mn_model.score(xb_train, yb_train)

0.6421267893660532

In [166]:
mn_model.score(xb_test, yb_test)

0.6219512195121951

Ordinal Logistic Regression

In [167]:
ord_model = LogisticRegression(solver = 'liblinear', class_weight={0:0,1:1})
ord_model.fit(xb_train, yb_train)

LogisticRegression(C=1.0, class_weight={0: 0, 1: 1}, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [168]:
ord_model.score(xb_train,yb_train)

0.5255623721881391

In [169]:
yb_pr = ord_model.predict(xb_train)

In [170]:
print(classification_report(yb_train, yb_pr))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00        29
           1       0.31      0.55      0.40        31
           2       0.44      0.40      0.42        35
           3       0.49      0.77      0.60        30
           4       0.45      0.15      0.23        33
           5       0.41      0.57      0.48        37
           6       0.64      0.50      0.56        36
           7       0.53      0.84      0.65        31
           8       0.89      0.50      0.64        32
           9       0.62      0.74      0.68        34
          10       0.61      0.53      0.57        32
          11       0.80      0.73      0.76        33
          12       0.00      0.00      0.00        30
          13       0.49      0.64      0.55        33
          14       0.53      0.91      0.67        33

    accuracy                           0.53       489
   macro avg       0.48      0.52      0.48       489
weighted avg       0.49   

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Changing Parameters

Penalty

In [171]:
bm = LogisticRegression(solver = 'liblinear', penalty= 'l2' , C=1.0)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)


0.7049180327868853

Dual

In [172]:
bm = LogisticRegression(solver = 'liblinear', penalty= 'l2' , dual = True)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


0.6721311475409836

max_iter

In [173]:
bm = LogisticRegression(max_iter=1000)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.6939890710382514

tol

In [174]:
bm = LogisticRegression(solver = 'liblinear', tol= 1e-5, C=1.0)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.7049180327868853

C

In [175]:
bm = LogisticRegression(tol= 1e-5, C=10.0)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.6939890710382514

fit_intecept

In [176]:
bm = LogisticRegression(solver = 'liblinear',tol = 1e-4, C=20.0, fit_intercept = False)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.6666666666666666

intercept_scaling

In [177]:
bm = LogisticRegression(solver = 'liblinear',penalty='l2',intercept_scaling=14.8)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.6939890710382514

class_weight

In [178]:
bm = LogisticRegression(solver = 'liblinear', class_weight={0:3.8,1:0.2})
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.32786885245901637

random_state

In [179]:
bm = LogisticRegression(solver = 'sag',random_state=58)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


0.6721311475409836

solver

In [180]:
bm = LogisticRegression(solver = 'saga')
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


0.6721311475409836

multi_class

In [181]:
bm = LogisticRegression(multi_class='ovr')
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.6939890710382514

verbose

In [182]:
bm = LogisticRegression(solver = 'liblinear', verbose=5)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

[LibLinear]

0.7049180327868853

l1_ratio

In [183]:
bm = LogisticRegression(solver='saga', penalty='elasticnet', l1_ratio=0.66)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


0.6721311475409836

n-jobs

In [184]:
bm = LogisticRegression(n_jobs=10)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.6939890710382514

warm_start

In [185]:
bm = LogisticRegression(solver='liblinear',warm_start= True)
bm.fit(x.values.reshape(-1,1),y)
bm.score(x.values.reshape(-1,1),y)

0.7049180327868853

# Inferance

Original Score(By Default Paramenters) = 0.6939890710382514

The following Scores was observed after changing parameters:

Penalty= l2

0.7049180327868853


Dual = True

0.6721311475409836



max_iter= 1000

0.6939890710382514



tol= 1e-5

0.7049180327868853



C=10

0.6939890710382514



fit_intecept=  False

0.6666666666666666



intercept_scaling = 14.8

0.6939890710382514



class_weight = 0:3.8,1:0.2

0.32786885245901637



random_state = 58

0.6721311475409836



solver= saga

0.6721311475409836



multi_class= ovr

0.6939890710382514



verbose=5

0.7049180327868853



l1_ratio=0.66

0.6721311475409836



n-jobs=10

0.6939890710382514



warm_start= True

0.7049180327868853



Thus improving penalty, tol , verbose and warm_start slightly increased the accuracy of the model(from about 69.4 to 70.5)

On other hand changes in l1_ratio, solver, class_weight, random_state and fit intercept decreased the accurcy of the model

While C, n1_jobs, , multi_class, incerpt_scaling and max_teir had no changes on accuracy
